virtual_rainforest/main.py

```diff
@@ -0,0 +1,125 @@
+"""Defines the function used to run a full simulation of the model.
+
+As well as setting up the function to run the overall virtual rainforest simulation,
+this script also defines the command line entry points for the model.
+"""
+
+from copy import deepcopy
+from typing import Any, Type, Union
+
+from virtual_rainforest.core.config import validate_config
+from virtual_rainforest.core.logger import LOGGER, log_and_raise
+from virtual_rainforest.core.model import (
+    MODEL_REGISTRY,
+    BaseModel,
+    FailedModel,
+    InitialisationError,
+)
+
+
+def select_models(model_list: list[str]) -> list[Type[BaseModel]]:
+    """Select the models to be run for a specific virtual rainforest simulation.
+
+    This function looks for models from a list of models, if these models can all be
+    found in the registry then they are returned. Otherwise an error is logged, which
+    should be handled appropriately downstream.
+
+    Args:
+        model_list: A list of models to select
+
+    Raises:
+        InitialisationError: If one or more models cannot be found in the registry
+    """
+
+    # Remove "core" from model list as it is not a model
+    if "core" in model_list:
+        model_list.remove("core")
+
+    LOGGER.info("Attempting to configure the following models: %s" % model_list)
+
+    # Make list of missing models, and return an error if necessary
+    miss_model = [model for model in model_list if model not in MODEL_REGISTRY.keys()]
+    if miss_model != []:
+        log_and_raise(
+            f"The following models cannot be configured as they are not found in the "
+            f"registry: {miss_model}",
+            InitialisationError,
+        )
+
+    # Then extract each model from the registry
+    modules = [MODEL_REGISTRY[model] for model in model_list]
+
+    return modules
+
+
+def configure_models(
+    config: dict[str, Any], model_list: list[Type[BaseModel]]
+) -> list[BaseModel]:
+    """Configure a set of models for use in a `virtual_rainforest` simulation.
+
+    Args:
+        config: The full virtual rainforest configuration
+        modules: A set of models to be configured
+
+    Raises:
+        InitialisationError: If one or more models cannot be properly configured
+    """
+
+    # Use factory methods to configure the desired models
+    models_cfd = [model.factory(config) for model in model_list]
+
+    if any([isinstance(model, FailedModel) for model in models_cfd]):
+        log_and_raise(
+            "Could not configure all the desired models, ending the simulation.",
+            InitialisationError,
+        )
```

Comment:
I think that a cleaner approach to do this without the `FailModel` class would be something like:

```suggestion
    failed_models = []
    models_cfd = []
    for model in model_list:
        try:
            models_cfd.append(model.factory(config))
        except (RelevantException, OrListOfThem):
            failed_models.append(model.name)

    if len(failed_models):
        log_and_raise(
            f"Could not configure all the desired models, ending the simulation. These failed {failed_models}.",
            InitialisationError,
        )
```
Which also shows the advantage of indicating which models failed.

Reply:
Yeah - much better. My brain was stuck in list comprehension mode, so wanting to end up with a list of objects that we could then scan for failure.